### ISSUE TITLE
Improve/optimize farming core logic loop

### ISSUE THREAD

**Issue by DenysAshikhin** [DenysAshikhin/gameplay-planner#18]
Desired result:

Instead of using arbitrary number of loops to jump x time in future and get estimated production amounts, use proper math to jump any amount at once and get exact values. 

This discussion is following the one between me and pg. The current version of the code is:

```

  const calc_future_state = function (plants, time) {

    let timeIncrease = 1; //how many seconds at a time we will skip forward
    let currentTime = 0; // how much time we have skipped forward in total

    let futurePlants = JSON.parse(JSON.stringify(plants)); // make a deep copy of plants to keep seperate from the passed copy and not edit it by accident
    let HP_produced = 0;

    function coolTimeThing(time, j) {
      let a = 1
      for (let i = 0; i < j; i++) {
        a *= time + i
      }
      return a
    }

    // keep looping until we run out of given time
    while (currentTime < time) {

      /** here I would have my logic to determine how far to skip in time, let's assume a constant 10seconds for now */
      timeIncrease = 10;

      //we can either loop from first plant to last, or opposite order, you can change this of course
      for (let i = 0; i < futurePlants.length; i++) {
        // when i=0, produce hp, when i>0 produce plant i
        let currentMultiplier = 1;
        // you probably are using somehting like Decimal.js to make numbers not cause issues
        // in which case replace 1 with whatever initates the object
        for (let j = 1; i + j <= futurePlants.length; j++) {
          currentMultiplier *= futurePlants[i - 1 + j].multiplier / j;
          if (i === 0) {
            HP_produced += futurePlants[i - 1 + j].amount * currentMultiplier * coolTimeThing(timeIncrease, j);
          } else { // 
            futurePlants[i - 1].amount += futurePlants[i - 1 + j].amount * currentMultiplier * coolTimeThing(timeIncrease, j);
          }
        }
      }

      // Update production values to reflect the new amounts
      for (let i = 0; i < futurePlants.length; i++) {
        futurePlants[i].production = futurePlants[i].amount * futurePlants[i].multiplier
      }
      currentTime += timeIncrease;
    }
    return { plants: futurePlants, HP_produced: HP_produced }
  }


  let plants_original = [{ amount: 1, production: 1, multiplier: 1 }, { amount: 1, production: 1, multiplier: 2 },];
  let result = calc_future_state(plants_original, 10);
  console.log(result);
```
We can also make stuff on jsfiddle to test
https://jsfiddle.net/thelastspark/ypsc5x26/1/

**Comment by DenysAshikhin** [comment]
Currently, we have an initial working (thanks pg) but we now need to take into account the multipliers increasing as the autoplots harvest plants.

The best I can offer atm is a function that given a plant, settings (such as time and number of autos), will return the future amount of harvests.

We can then use a different function to get the multiplier at the end. Then we can combine the initial production (where production is a combination of multiplier and number of that plant we have) and the end production. We *can* reason a linear increase here but that would be wrong. While the multiplier would increase linearly, since autoplots harvest at a constant rate (we are simplifying things like plant growth timers going down for now); the amount of plant we have is not linear. As the plant before it is increase as time goes on, so each tick produces more of the current plant.

Hope that makes sense lol. 
Basically we need to handle a growing amount + multiplier in an elegant fashion next!! :heart:
